source/utilities.cc

```diff
@@ -1822,6 +1822,188 @@ namespace aspect
       return lookup->get_data(position,component);
     }
 
+
+
+    double
+    weighted_p_norm_average ( const std::vector<double> &weights,
+                              const std::vector<double> &parameter_values,
+                              const double p)
+    {
+      //TODO: prevent devision by zero for all
+      double averaged_parameter = 0.0;
+
+      // first look at the specal case which can be done faster
+      // Minimum
+      if (p <= -1000)
+        {
+          const unsigned int i = (unsigned int)(std::min_element( weights.begin(),
+                                                                  weights.end() )
+                                                - weights.begin());
+          averaged_parameter = parameter_values[i];
+          return averaged_parameter;
+        }
+      // Harmonic average (TODO: Check if we need to use a epsilon)
+      if (p == -1)
+        {
+          for (unsigned int i=0; i< weights.size(); ++i)
+            {
+              if (parameter_values[i] == 0 && weights[i] != 0)
+                return 0;
+              else if (parameter_values[i] !=0)
+                averaged_parameter += weights[i]/(parameter_values[i]);
+            }
+          averaged_parameter = 1.0/averaged_parameter;
+          return averaged_parameter;
+        }
+
+      // Geometric average (TODO: Check if we need to use a epsilon)
+      if (p == 0)
+        {
+          for (unsigned int i=0; i < weights.size(); ++i)
+            averaged_parameter += weights[i]*std::log(parameter_values[i]);
+          averaged_parameter = std::exp(averaged_parameter);
+          return averaged_parameter;
+        }
+
+      // Arithmetic average (TODO: Check if we need to use a epsilon)
+      if (p == 1)
+        {
+          for (unsigned int i=0; i< weights.size(); ++i)
+            averaged_parameter += weights[i]*parameter_values[i];
+          return averaged_parameter;
+        }
+
+      // Quadratic average (RMS) (TODO: Check if we need to use a epsilon)
+      if (p == 2)
+        {
+          for (unsigned int i=0; i< weights.size(); ++i)
+            averaged_parameter += weights[i]*parameter_values[i]*parameter_values[i];
+          averaged_parameter = std::sqrt(averaged_parameter);
+          return averaged_parameter;
```

Comment:
just say `return std::sqrt(...)` in the previous line, instead of re-using the variable.